[BUG]客戶端崩潰

系統：
Architecture	ARMv7 Processor rev 1 (v7l)
Firmware Version	OpenWrt 19.07.7 r11306-c4a6851c72

開啟ｒｏｕｔｅｒ的情況下直連訪問不會報錯，使用代理則崩潰：

[INFO]  2021/06/02 01:37:34 trojan-go v0.10.3 initializing
panic: unaligned 64-bit atomic operation

goroutine 16 [running]:
runtime/internal/atomic.panicUnaligned()
        runtime/internal/atomic/unaligned.go:8 +0x24
runtime/internal/atomic.Xadd64(0x2d8f71c, 0x8, 0x0, 0x8, 0x8)
        runtime/internal/atomic/asm_arm.s:233 +0x14
github.com/p4gefau1t/trojan-go/tunnel/trojan.(*OutboundConn).Read(0x2d8f710, 0x2d08000, 0x8, 0x8, 0x1a054, 0xb66dd0, 0x0)
        github.com/p4gefau1t/trojan-go/tunnel/trojan/client.go:83 +0x9c
io.ReadAtLeast(0xa6c86010, 0x2d9e7c0, 0x2d08000, 0x8, 0x8, 0x8, 0x0, 0x0, 0x0)
        io/io.go:328 +0x6c
io.ReadFull(...)
        io/io.go:347
github.com/xtaci/smux.(*Session).recvLoop(0x2bf3360)
        github.com/xtaci/smux@v1.5.15/session.go:320 +0x128
created by github.com/xtaci/smux.newSession
        github.com/xtaci/smux@v1.5.15/session.go:105 +0x240



客戶端配置：
{
    "run_type": "client",
    "local_addr": "127.0.0.1",
    "local_port": 1081,
    "remote_addr": "",
    "remote_port": 443,
    "log_level": 1,
    "log_file": "/var/log/trojan-go-error.log",
    "password": [
        ""
    ],
    "ssl": {
        "sni": ""
    },
    "mux": {
        "enabled": true,
        "concurrency": 8,
        "idle_timeout": 60
    },
    "router": {
        "enabled": true,
        "bypass": [
            "geoip:cn",
            "geoip:private",
            "geosite:cn",
            "geosite:private"
        ],
        "block": [
            "geosite:category-ads"
        ],
        "proxy": [
            "geosite:geolocation-!cn"
        ],
        "default_policy": "proxy",
        "domain_strategy": "as_is",
        "geoip": "/mnt/sda1/bin/trojan-go/geoip.dat",
        "geosite": "/mnt/sda1/bin/trojan-go/geosite.dat"
    },
    "websocket": {
        "enabled": true,
        "path": "",
        "host": ""
    },
    "shadowsocks": {
        "enabled": true,
        "method": "CHACHA20-IETF-POLY1305",
        "password": ""
    }
}


版本v0.10.3崩潰　，v0.10.0正常

編譯環境：go version go1.16.4 windows/amd64：

panic: unaligned 64-bit atomic operation

goroutine 26 [running]:
runtime/internal/atomic.panicUnaligned()
        C:/Program Files/Go/src/runtime/internal/atomic/unaligned.go:8 +0x24
runtime/internal/atomic.Xadd64(0x27b371c, 0x8, 0x0, 0x8, 0x8)
        C:/Program Files/Go/src/runtime/internal/atomic/asm_arm.s:233 +0x14
github.com/p4gefau1t/trojan-go/tunnel/trojan.(*OutboundConn).Read(0x27b3710, 0x2858bd8, 0x8, 0x8, 0x1a054, 0xb76dd0, 0x0)
        C:/Users/costom/Desktop/trojan-go-0.10.3/tunnel/trojan/client.go:83 +0x9c
io.ReadAtLeast(0xa6c23960, 0x28567f0, 0x2858bd8, 0x8, 0x8, 0x8, 0x0, 0x0, 0x0)
        C:/Program Files/Go/src/io/io.go:328 +0x6c
io.ReadFull(...)
        C:/Program Files/Go/src/io/io.go:347
github.com/xtaci/smux.(*Session).recvLoop(0x271b360)
        C:/Users/costom/go/pkg/mod/github.com/xtaci/smux@v1.5.15/session.go:320 +0x128
created by github.com/xtaci/smux.newSession
        C:/Users/costom/go/pkg/mod/github.com/xtaci/smux@v1.5.15/session.go:105 +0x240

--- a/statistic/memory/memory.go
+++ b/statistic/memory/memory.go
@@ -17,12 +17,18 @@ import (
 const Name = "MEMORY"
 
 type User struct {
-	sent        uint64
-	recv        uint64
-	lastSent    uint64
-	lastRecv    uint64
-	sendSpeed   uint64
-	recvSpeed   uint64
+	// WARNING: do not change the order of these fields.
+	// 64-bit fields that use `sync/atomic` package functions
+	// must be 64-bit aligned on 32-bit systems.
+	// Reference: https://github.com/golang/go/issues/599
+	// Solution: https://github.com/golang/go/issues/11891#issuecomment-433623786
+	sent      uint64
+	recv      uint64
+	lastSent  uint64
+	lastRecv  uint64
+	sendSpeed uint64
+	recvSpeed uint64
+
 	hash        string
 	ipTable     sync.Map
 	ipNum       int32

--- a/tunnel/trojan/client.go
+++ b/tunnel/trojan/client.go
@@ -29,9 +29,15 @@ const (
 )
 
 type OutboundConn struct {
+	// WARNING: do not change the order of these fields.
+	// 64-bit fields that use `sync/atomic` package functions
+	// must be 64-bit aligned on 32-bit systems.
+	// Reference: https://github.com/golang/go/issues/599
+	// Solution: https://github.com/golang/go/issues/11891#issuecomment-433623786
+	sent uint64
+	recv uint64
+
 	metadata          *tunnel.Metadata
-	sent              uint64
-	recv              uint64
 	user              statistic.User
 	headerWrittenOnce sync.Once
 	net.Conn

--- a/tunnel/trojan/server.go
+++ b/tunnel/trojan/server.go
@@ -21,9 +21,15 @@ import (
 
 // InboundConn is a trojan inbound connection
 type InboundConn struct {
+	// WARNING: do not change the order of these fields.
+	// 64-bit fields that use `sync/atomic` package functions
+	// must be 64-bit aligned on 32-bit systems.
+	// Reference: https://github.com/golang/go/issues/599
+	// Solution: https://github.com/golang/go/issues/11891#issuecomment-433623786
+	sent uint64
+	recv uint64
+
 	net.Conn
-	sent     uint64
-	recv     uint64
 	auth     statistic.Authenticator
 	user     statistic.User
 	hash     string

